src/mscorlib/src/System/Text/Encoding.cs

```diff
@@ -867,23 +867,36 @@ public virtual int GetByteCount(char[] chars)
         }
 
         [Pure]
-        public virtual int GetByteCount(String s)
-        {
-            if (s==null)
-                throw new ArgumentNullException(nameof(s));
-            Contract.EndContractBlock();
-
-            char[] chars = s.ToCharArray();
-            return GetByteCount(chars, 0, chars.Length);
-
-        }
+        public virtual int GetByteCount(String s) => GetByteCount(s, 0, s.Length);
 
         // Returns the number of bytes required to encode a range of characters in
         // a character array.
         //
         [Pure]
         public abstract int GetByteCount(char[] chars, int index, int count);
 
+        // Returns the number of bytes required to encode a string range.
+        //
+        [Pure]
+        public virtual int GetByteCount(string s, int index, int count)
+        {
+            if (s == null)
+                throw new ArgumentNullException(nameof(s), 
+                    Environment.GetResourceString("ArgumentNull_String"));
+            if (index < 0)
+                throw new ArgumentOutOfRangeException(nameof(index),
+                      Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+            if (count < 0)
+                throw new ArgumentOutOfRangeException(nameof(count),
+                      Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+            if (index > s.Length - count)
+                throw new ArgumentOutOfRangeException(nameof(index),
+                      Environment.GetResourceString("ArgumentOutOfRange_IndexCount"));
+            Contract.EndContractBlock();
+
+            return s.GetByteCountFromEncoding(index, count, this);
```

Comment:
>s.GetByteCountFromEncoding(index, count, this); [](start = 19, length = 47)

this should something like

fixed (char *pChar = s)
{
    return GetByteCount(pChar+index, count);
}

and do the same in GetBytes